src/freenet/support/PrioritizedTicker.java

```diff
@@ -267,54 +250,81 @@ public Executor getExecutor() {
 
 	public int queuedJobs() {
 		synchronized(timedJobsByTime) {
-			return timedJobsByTime.size();
+			return timedJobsQueued.size();
 		}
 	}
 
+    public int queuedJobsUniqueTimes() {
+        synchronized(timedJobsByTime) {
+            return timedJobsByTime.size();
+        }
+    }
+
 	@Override
+	/* Remove a queued job.
+	 * @param runnable The job to remove. If this is currently queued, it will be 
+	 * removed. The Ticker should not throw if the job is not queued. */
 	public void removeQueuedJob(Runnable runnable) {
 		Job job = new Job(null, runnable);
 		synchronized(timedJobsByTime) {
 			Long t = timedJobsQueued.remove(job);
 			if(t != null) {
-				Object o = timedJobsByTime.get(t);
-				if(o == null) return; // XXX impossible -> assert
-				if(o instanceof Job) {
-					assert(o.equals(job));
-					timedJobsByTime.remove(t);
-				} else {
-					Job[] jobs = (Job[]) o;
-					if(jobs.length == 1) {
-						assert(jobs[0].equals(job));
-						timedJobsByTime.remove(t);
-					} else {
-						Job[] newJobs = new Job[jobs.length-1];
-						int x = 0;
-						for(Job oldjob : jobs) {
-							if(oldjob.equals(job)) {
-								continue;
-							}
-							newJobs[x++] = oldjob;
-							if(x == jobs.length) {
-								assert(false);
-								newJobs = jobs;
-							}
-						}
-						if(x == 0) {
-							assert(false);
-							timedJobsByTime.remove(t);
-						} else if (x == 1) {
-							timedJobsByTime.put(t, newJobs[0]);
-						} else {
-							if(x != newJobs.length)
-								newJobs = Arrays.copyOf(newJobs, x);
-							timedJobsByTime.put(t, newJobs);
-							assert(x == jobs.length-1);
-						}
-					}
-				}
+			    removeQueuedJobInner(job, t);
 			}
 		}
 	}
-	
+
+	/** Remove a queued job from the internal structures other than timedJobsQueued. The
+	 * caller is responsible for that.
+	 * @param job The job to remove.
+	 * @param t The time at which is it scheduled.
+	 */
+	private void removeQueuedJobInner(Job job, Long t) {
+        Object o = timedJobsByTime.get(t);
+        if(o == null) {
+            Logger.error(this, "Job in timedJobsQueued but not in timedJobsByTime");
+            // Cleaned up so no reason to throw in this critical structure.
```

Comment:
Elseriseth [has opinions](SSK%40WPECGLtYbVi8Rl6Y7Vzl2Lvd2EUVW99v3yNV3IWROG8%2CMnwtfrASwEUVgXVCxNALLwCKqe6woGt3biDRVcBulkk%2CAQACAAE/fms|2015-08-17|Message-1) on this:

> GRRR!
> That's outright insane.
> Those asserts should be rather removed, than converted to
> Logger.error.
> If condition MUST NEVER EVER HAPPEN, it should be either ignored, or be
> HARD ERROR, not just one more random message. And released (non-debug,
> post-alpha) binary SHOULD NOT EVEN CHECK for those condition (by
> disabling asserts).

I agree that if the data structure has a bug we should have it fail loudly so we can fix it instead of adding another log message to the fire hose. Without benchmarks showing that asserts are a bottleneck I disagree that asserts should be disabled in release mode.